### PR TITLE
feat(platform/bitbucket): autodiscoverProjects

### LINKED ISSUE #27661: autodiscoverProjects for Bitbucket Cloud
### Describe the proposed change(s).

Discussed in https://github.com/renovatebot/renovate/discussions/27093

I think it's best to make this a new option because, apart from it being a different name (projects vs namespaces), Bitbucket Cloud also has these differences vs GitLab:
- Filtering would be done client-side not server-side
- Ability to use patterns and/or excludes

So we should create a new option autodiscoverProjects, for Bitbucket Cloud only, and use the `matchRegexOrGlobList()` function which allows exclusions.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Add new config option `autodiscoverProjects` for the bitbucket platform which will filter the list of autodiscovered repositories 
- It can be used to both include/exclude projects 

## Context
Closes: https://github.com/renovatebot/renovate/issues/27661

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (tested locally)

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
Unfortunately, tokens which provide access to all projects is only avaiable to premium users so I tested using repository access tokens. In the two tests I performed, I used a repo token to fetch the said test-repo and then use `autodiscoverProjects` filter to  once allow the project the repo belonged to and once disallow, I have pasted the logs below for both cases:

Test repo name: rahugautamsingh/ignore/test-repo

**When project: ignore was allowed**
```
DEBUG: Combined config
       "config": {
         "forkProcessing": true,
         "token": "***********",
         "endpoint": "https://api.bitbucket.org/",
         "platform": "bitbucket",
         "autodiscover": true,
         "hostRules": [{"matchHost": "https://api.github.com", "token": "***********"}]
       }
       ---
DEBUG: Autodiscovering Bitbucket Cloud repositories
DEBUG: hostRules: applying Bearer authentication for api.bitbucket.org
DEBUG: Autodiscovered 1 repositories
 INFO: Autodiscovered repositories
       "length": 1,
       "repositories": ["rahulgautamsingh/test-repo"]
```

When the project wasn't allowed
```
DEBUG: Combined config
       "config": {
         "forkProcessing": true,
         "token": "***********",
         "endpoint": "https://api.bitbucket.org/",
         "platform": "bitbucket",
         "autodiscover": true,
         "autodiscoverProjects": ["!ignore"],
         "hostRules": [{"matchHost": "https://api.github.com", "token": "***********"}]
       }
       ---
DEBUG: Applying autodiscoverProjects filter
       "autodiscoverProjects": ["!ignore"]
DEBUG: No repositories were autodiscovered
```

**Comment by rarkins** [comment]
@FHannes @setchy for your review if you have time

File: docs/usage/self-hosted-configuration.md [hunk 1]
```diff
@@ -213,6 +213,22 @@ For example:
...
+For example:
+
+```json
+{
+  "platform": "bitbucket",
+  "autodiscoverProjects": ["a-group", "!another-group/some-subgroup"]
+}
+```
+
+The `autodiscoverProjects` config option takes an array of minimatch-compatible globs or re2-compatible regex strings.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:229]:
```suggestion
The `autodiscoverProjects` config option takes an array of minimatch-compatible globs or RE2-compatible regex strings.
```

File: docs/usage/self-hosted-configuration.md [hunk 2]
```diff
@@ -213,6 +213,22 @@ For example:
 }
 ```
 
+## autodiscoverProjects
+
+You can use this option to filter the list of autodiscovered repositories by project names.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:218]: I would like a extra line explaining _why_ this can be handy, and/or _what_ the intended use is.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by HonkingGoose — APPROVED** [review]

File: lib/modules/platform/bitbucket/index.ts [hunk 3]
```diff
@@ -113,17 +114,30 @@ export async function initPlatform({
...
+
+    // if autodiscoverProjects is configured
+    // filter the repos list
+    if (config?.projects?.length) {
+      logger.debug(
+        { autodiscoverProjects: config?.projects },
+        'Applying autodiscoverProjects filter',
+      );
+      repos = repos.filter((repo) =>
+        matchRegexOrGlobList(repo.project.name, config.projects!),
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:137]:
```suggestion
        matchRegexOrGlobList(repo.project.name, config.projects),
```
you already validate above

File: lib/modules/platform/bitbucket/index.ts [hunk 4]
```diff
@@ -113,17 +114,30 @@ export async function initPlatform({
 }
 
 // Get all repositories that the user has access to
-export async function getRepos(): Promise<string[]> {
+export async function getRepos(config?: AutodiscoverConfig): Promise<string[]> {
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:117]:
```suggestion
export async function getRepos(config: AutodiscoverConfig): Promise<string[]> {
```
I think it's not optional and will always be passed from worker, so needs test updates.

File: lib/modules/platform/bitbucket/index.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/modules/platform/bitbucket/index.ts:137]: TS throws errors for some reason :)

**Comment by FHannes** [comment]
@rarkins The implementation looks to meet my requirements for the feature. Given there's no artifacts for this and I deploy Renovate as a container, it's a bit tricky for me to actually test the functionality at this time though.

**Review by viceice — APPROVED** [review]

**Comment by FHannes** [comment]
I was able to test this and it looks to be working as described.

**Pull request merged**